scp/cc/cpio/client_providers/private_key_fetcher_provider/src/azure/azure_private_key_fetcher_provider_utils.cc

```diff
@@ -29,16 +28,305 @@ using google::scp::core::Uri;
   CHECK(report.has_value()) << "Failed to get attestation report";
-  http_request.body = core::BytesBuffer(nlohmann::json(report.value()).dump());
+  json_obj["attestation"] = report.value();
+
+  http_request.body = core::BytesBuffer(json_obj.dump());
+  return privateKey;
+}
+
+/**
+ * @brief Generate a new wrapping key
+ */
+std::pair<EVP_PKEY*, EVP_PKEY*>
+AzurePrivateKeyFetchingClientUtils::GenerateWrappingKey() {
+  RSA* rsa = RSA_new();
+  BIGNUM* e = BN_new();
+
+  BN_set_word(e, RSA_F4);
+  RSA_generate_key_ex(rsa, 4096, e, NULL);
+
+  EVP_PKEY* private_key = EVP_PKEY_new();
+  if (private_key == NULL) {
+    char* error_string = ERR_error_string(ERR_get_error(), NULL);
+    throw std::runtime_error(std::string("New EVP_PKEY failed: ") +
+                             error_string);
+  }
+
+  if (EVP_PKEY_set1_RSA(private_key, rsa) != 1) {
+    char* error_string = ERR_error_string(ERR_get_error(), NULL);
+    throw std::runtime_error(std::string("Set RSA key failed: ") +
+                             error_string);
+  }
+
+  // Create a new EVP_PKEY for the public key
+  EVP_PKEY* public_key = EVP_PKEY_new();
+  if (public_key == NULL) {
+    char* error_string = ERR_error_string(ERR_get_error(), NULL);
+    throw std::runtime_error(std::string("New EVP_PKEY (public) failed: ") +
+                             error_string);
+  }
+
+  // Get the RSA public key structure
+  const RSA* rsa_pub = EVP_PKEY_get1_RSA(private_key);
+  if (rsa_pub == NULL) {
+    char* error_string = ERR_error_string(ERR_get_error(), NULL);
+    throw std::runtime_error(std::string("Get RSA public key failed: ") +
+                             error_string);
+  }
+
+  // Duplicate the RSA public key structure
+  RSA* rsa_pub_dup = RSA_new();
+  if (rsa_pub_dup == NULL) {
+    char* error_string = ERR_error_string(ERR_get_error(), NULL);
+    throw std::runtime_error(
+        std::string("Create RSA public key duplicate failed: ") + error_string);
+  }
+
+  if (!RSA_set0_key(rsa_pub_dup, rsa_pub->n, rsa_pub->e, NULL)) {
+    char* error_string = ERR_error_string(ERR_get_error(), NULL);
+    throw std::runtime_error(
+        std::string("Set RSA public key duplicate values failed: ") +
+        error_string);
+  }
+
+  RSA_up_ref(rsa_pub_dup);
+
+  // Set the duplicated RSA public key structure to the public_key EVP_PKEY
+  if (EVP_PKEY_set1_RSA(public_key, rsa_pub_dup) != 1) {
+    char* error_string = ERR_error_string(ERR_get_error(), NULL);
+    throw std::runtime_error(std::string("Set RSA public key failed: ") +
+                             error_string);
+  }
+
+  BN_free(e);
+  RSA_free(rsa);  // Free the RSA structure if we're done with it
+
+  return std::make_pair(private_key, public_key);
+}
+
+/**
+ * @brief Convert a wrapping key in PEM
+ *
+ * @param wrappingKey RSA public key used to wrap a key.
+ */
+std::string AzurePrivateKeyFetchingClientUtils::EvpPkeyToPem(EVP_PKEY* pkey) {
+  BIO* bio = BIO_new(BIO_s_mem());
```

Comment:
Can you use [RAII ](https://en.cppreference.com/w/cpp/language/raii) style wrapper for these OpenSSL functions please (to avoid memory leak-ish errors)?

Reply:
done